android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC.java

```diff
@@ -272,9 +275,13 @@ private SecretKey tryGenerateStrongBoxSecurityKey(String service) throws NoSuchA
         }
         try {
             return generateKey(getKeyGenSpecBuilder(service).setIsStrongBoxBacked(true).build());
-        } catch (StrongBoxUnavailableException e) {
+        } catch (Exception e) {
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && e instanceof StrongBoxUnavailableException) {
```

Comment:
why are you checking `Build.VERSION.SDK_INT >= Build.VERSION_CODES.P` if there is

```
if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
            return null;
        }
```
above?

Reply:
That was pre-existing code that I didn't touch. I did check just now on an API 23 emulator without that check. Without it the app crashes with a NoSuchMethodError.

The change I made is to satisfy the Verifier which doesn't seem to be intelligent enough (at least in API 19) to recognize that the earlier return means the StrongBoxUnavailableException will never be hit. Again, I'm not an Android expert but this is a pattern I saw others try to get the Verifier to stop having an issue resolving the exception. Issue https://github.com/oblador/react-native-keychain/issues/206 shows the crash that's resulting from this not being there although the things highlighted in red aren't the issue. It's that StrongBoxUnavailableException can't be resolved.

Reply:
```
if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
     return null;
}
```

means that the function will return null on api levels below `P` and so the check in 

`if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && e instanceof StrongBoxUnavailableException)`

does not have to be there at all.

if you're saying that without it, the app crashes with a NoSuchMethodError, then there is something else going on. Can you please double check what the problem is? I also might have time to take a look at it this week, perhaps I'll be able to shed some more light on this.

Reply:
Logically yes, it makes sense and I know what you're saying. The problem is this Java Verifier doesn't seem to be as smart.

I don't have the text with me, but this comment has an image of the crash:
https://github.com/oblador/react-native-keychain/issues/206#issue-427657210
Where the critical part from what I figured out is the "VFY unable to resolve exception class 466 (Landroid/security/keystore/StrongBoxUnavailableException)" error. Everything else seems to get replaced or resolved or whatever but this doesn't.

Reply:
@astapinski but this should only depend on the compile-time enviromnent and not runtime. If you build with the latest Android SDK, you should have this exception there.

Reply:
Maybe it'll help if I create a repro, I've done so here:
https://github.com/astapinski/rn_keychain_android_4_crash

Like the original poster did, I created a clean RN 0.57.8 project and added 3.1.1 in. When one first tries that, it fails with missing symbols/classes because it's using compileSdkVersion 27 (missing Build.VERSION_CODES.P, similar things). I set that to 28 in the repro and it then compiles but crashes at runtime with the same error they encountered, mainly:

`W/dalvikvm( 2723): VFY: unable to resolve exception class 356 (Landroid/security/keystore/StrongBoxUnavailableException;)
W/dalvikvm( 2723): VFY: unable to find exception handler at addr 0x1a
W/dalvikvm( 2723): VFY:  rejected Lcom/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC;.tryGenerateStrongBoxSecurityKey (Ljava/lang/String;)Ljavax/crypto/SecretKey;
W/dalvikvm( 2723): VFY:  rejecting opcode 0x0d at 0x001a
W/dalvikvm( 2723): VFY:  rejected Lcom/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC;.tryGenerateStrongBoxSecurityKey (Ljava/lang/String;)Ljavax/crypto/SecretKey;
W/dalvikvm( 2723): Verifier rejected class Lcom/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC;
W/dalvikvm( 2723): threadid=19: thread exiting with uncaught exception (group=0x9cd89b20)
D/gralloc_ranchu( 1142): gralloc_alloc: Creating ashmem region of size 1323008
E/AndroidRuntime( 2723): FATAL EXCEPTION: Thread-108
E/AndroidRuntime( 2723): Process: com.rn_keychain_android_4_crash, PID: 2723
E/AndroidRuntime( 2723): java.lang.VerifyError: com/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC
`
So yeah at compile time with 27 things are failed to be found. Go to 28 and things compile but it does fail at runtime. I did try the build tools 28.0.3 and support lib 28.0.0 as well, but it didn't change anything.

Reply:
I spent more time on it this morning since I was curious why the Verifier cared about a build check. It doesn't, the change to catch(Exception) and instanceof is all that was needed to stop the crash. I've removed the build version check and pushed, take a look now.